Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/Migration/Tools/MigrationWindow.cs

```diff
@@ -0,0 +1,417 @@
+                var migrationHandlerType = migrationTypesMap[selectedMigrationTypeName];
+                return Activator.CreateInstance(migrationHandlerType) as IMigrationHandler;
+            }
+        }
+
+        [MenuItem("Mixed Reality Toolkit/Utilities/Migration Window", false, 4)]
+        private static void ShowWindow()
+        {
+            var window = GetWindow<MigrationWindow>(typeof(SceneView));
+            window.titleContent = new GUIContent(WindowTitle, EditorGUIUtility.IconContent("d_TimelineEditModeRippleOFF").image);
+            window.minSize = new Vector2(400.0f, 300.0f);
+            window.Show();
+        }
+
+        private void OnGUI()
+        {
+            if (EditorApplication.isPlaying || EditorApplication.isPaused)
+            {
+                GUI.enabled = false;
+            }
+            DrawHeader();
+            DrawMigrationTypeSelector();
+            DrawMigrationToolbars();
+        }
+
+        private void OnFocus()
+        {
+            RefreshAvailableTypes();
+        }
+
+        private void RefreshAvailableTypes()
+        {
+            var type = typeof(IMigrationHandler);
+            var types = AppDomain.CurrentDomain.GetAssemblies()
+                .SelectMany(x => x.GetTypes())
+                .Where(x => type.IsAssignableFrom(x) && !x.IsInterface && !x.IsAbstract).ToList();
+
+            foreach (var migrationHandlerType in types)
+            {
+                if (!migrationTypesMap.ContainsKey(migrationHandlerType.Name))
+                {
+                    migrationTypesMap.Add(migrationHandlerType.Name, migrationHandlerType);
+                }
+            }
+            migrationTypeNames = new String[migrationTypesMap.Count + 1];
+
+            for (int i = 0; i < migrationTypesMap.Count; i++)
+            {
+                migrationTypeNames[i + 1] = migrationTypesMap.ElementAt(i).Key;
+            }
+        }
+
+        private void DrawHeader()
+        {
+            MixedRealityInspectorUtility.RenderMixedRealityToolkitLogo();
+
+            using (new EditorGUILayout.HorizontalScope())
+            {
+                EditorGUILayout.LabelField("Migration Window", EditorStyles.boldLabel);
+                InspectorUIUtility.RenderDocumentationButton(DependencyWindowURL);
+            }
+            EditorGUILayout.LabelField(WindowDescription, EditorStyles.wordWrappedLabel);
+
+            EditorGUILayout.Space();
+            EditorGUILayout.Space();
+        }
+
+        private void DrawMigrationTypeSelector()
+        {
+            using (var horizontal = new EditorGUILayout.HorizontalScope())
+            {
+                EditorGUILayout.LabelField("Migration Handler Selection", EditorStyles.boldLabel);
+                selectedMigrationHandlerIndex = EditorGUILayout.Popup(selectedMigrationHandlerIndex, migrationTypeNames, GUILayout.Width(400));
+            }
+
+            if (selectedMigrationHandlerIndex == 0)
```

Comment:
Calling this out not because I think you should change this, but because there's a general gotcha associated with this type of code.

DrawMigrationTypeSelector is called before DrawMigrationToolbars and the order here is important because GUI.enabled = false is called here. There's an implicit dependency between these two functions that is order (i.e. toolbars depends on type selector to have been called first, and this isn't called out anywhere in comments).

The danger here is people reorganizing stuff in the future and leaving part of the GUI in an enabled state, but the rest of it disabled, and there's no context for future folks to understand that dependency.

I think that because of the relative small scope of things here I don't think it really matters, but it's a thing to keep in mind for future changes.

Reply:
agree. adding comment.

Reply:
removed dependency